src/Symfony/Bundle/FrameworkBundle/Routing/Router.php

```diff
@@ -32,22 +33,29 @@ class Router extends BaseRouter implements WarmableInterface, ServiceSubscriberI
 {
     private $container;
     private $collectedParameters = array();
+    private $paramFetcher;
 
     /**
-     * @param ContainerInterface   $container A ContainerInterface instance
-     * @param mixed                $resource  The main resource to load
-     * @param array                $options   An array of options
-     * @param RequestContext       $context   The context
-     * @param LoggerInterface|null $logger
+     * @param ContainerInterface      $container  A ContainerInterface instance
+     * @param mixed                   $resource   The main resource to load
+     * @param array                   $options    An array of options
+     * @param RequestContext          $context    The context
+     * @param ContainerInterface|null $parameters A ContainerInterface instance allowing to fetch parameters
+     * @param LoggerInterface|null    $logger
      */
-    public function __construct(ContainerInterface $container, $resource, array $options = array(), RequestContext $context = null, LoggerInterface $logger = null)
+    public function __construct(ContainerInterface $container, $resource, array $options = array(), RequestContext $context = null, ContainerInterface $parameters = null, LoggerInterface $logger = null)
     {
         $this->container = $container;
-
         $this->resource = $resource;
         $this->context = $context ?: new RequestContext();
         $this->logger = $logger;
         $this->setOptions($options);
+
+        if ($parameters) {
+            $this->paramFetcher = array($parameters, 'get');
+        } elseif ($container instanceof SymfonyContainerInterface) {
+            $this->paramFetcher = array($container, 'getParameter');
```

Comment:
Could be simplified by using `new ContainerBag($container)` and always use `$this->parameters->get()` instead of this callable `$paramFetcher` property, but I'd be the one breaking the 3-cross-4 compat' 😄